### PR TITLE
Add In-Reply-To, References headers to email

### PR DESCRIPTION
**Pull request by mmd-osm**
Closes https://github.com/openstreetmap/openstreetmap-website/issues/2430

Context: _I'm subscribed to a blog post/changeset discussion/note and receive lots of emails by users replying to that post. As those emails don't include an In-Reply-To & References header, my inbox fills up with pages and pages of unrelated messages, rather than producing a nice threaded view of all emails._

### CODE REVIEW

File: app/mailers/notifier.rb [hunk 1]
```diff
@@ -94,11 +94,14 @@ def diary_comment_notification(comment, recipient)
...
       @commenturl = diary_entry_url(comment.diary_entry.user, comment.diary_entry, :anchor => "newcomment")
       @replyurl = new_message_url(comment.user, :message => { :title => "Re: #{comment.diary_entry.title}" })
-
+      @ref = "osm-diary-#{comment.diary_entry.id}@#{Settings.server_url}"
       @author = @from_user
 
       attach_user_avatar(comment.user)
 
+      headers["In-Reply-To"] = @ref
+      headers["References"]  = @ref
```
Comment (tomhughes) [inline app/mailers/notifier.rb]: I think we could reduce the amount of duplication here by having a `set_references` helper function and then just doing `set_references("diary", comment.diary_entry)` etc in each of the three methods?

Comment (mmd-osm) [reply app/mailers/notifier.rb]: Good point, added a new helper method now.

**Comment by mmd-osm** [comment]
I tested this change  on master.apis.dev now. Results look good in Thunderbird: comments to a note from different people, close, reactive, are all in one thread, in they share the same In-Reply-To header.

However, Google Mail now starts merging emails that share the same subject, although the HTTP References header differs. Even worse, different types of messages are now grouped, even they belong to different objects.

I think one way of addressing this would be to use less generic email title (see my comment here: https://github.com/openstreetmap/openstreetmap-website/issues/2430#issuecomment-552435433).

**Comment by tomhughes** [comment]
That sounds very odd... I thought gmail used the Message-ID and any references to link threads.

If anything you'd expect it to fall back to title matching where there is no reference information and adding that ought to improve it!

**Comment by tomhughes** [comment]
Actually is that the problem, that there is no root message with that ID for it to anchor the thread on?

**Comment by mmd-osm** [comment]
The Github notification mails I receive for this pull request don't have a root message, only those two header fields. However, all mails share the same title (e.g.` [openstreetmap/openstreetmap-website] Add In-Reply-To, References headers to email (#2462)`), and that may be the reason why threading works with Github notifications in gmail. Current osm emails don't include anything specific except for the user name that sent the email maybe.

I'm a bit reluctant to push this change as it is now to production, given that probably a fair amount of people uses gmail inbox. Can we somehow refine this on master for a bit without pushing to production, or have it inactive in production?

**Comment by tomhughes** [comment]
I haven't pushed it to production yet.

**Comment by mmd-osm** [comment]
I think https://gsuiteupdates.googleblog.com/2019/03/threading-changes-in-gmail-conversation-view.html —> additional details would be relevant here: besides the header, we need to make sure that messages that should belong together have the same email subject.

**Comment by mmd-osm** [comment]
Before changing any of the existing messages (which is a lot of work), I'd like to try two other things:

* Add new header `X-Entity-Ref-ID`  -> set value to `ref`, check if this fixes threading
* Add new header `X-Entity-Ref-ID`  -> set value to some random value to turn off threading in gmail altogether

**Comment by tomhughes** [comment]
I have https://tomh.apis.dev.openstreetmap.org/ running with `X-Entity-Ref-ID` set to the `ref` value and that seems to stop gmail threading them at all...

**Comment by mmd-osm** [comment]
Thanks! Indeed, this turns threading off in gmail, which I think is ok for the time being.

Assuming we change some of the wording of those messages, we could remove the `X-Entity-Ref-ID `header at a later point in time. In particular there's no need to include `%{user}` in the email subject, as we already have that very same information in the "From:" header (at least on production).

**Comment by tomhughes** [comment]
I've merged that to master now. Doesn't this mean we're just back where we started though?!?

**Comment by tomhughes** [comment]
I guess it may thread for other mail clients.

**Comment by mmd-osm** [comment]
Exactly. I just checked Thunderbird, and it's still threading as expected.

**Pull request merged**